File: app/src/main/java/com/jawnnypoo/bleh/service/MehClient.java
package com.jawnnypoo.bleh.service;

import com.jawnnypoo.bleh.BuildConfig;

import retrofit.Callback;
import retrofit.RequestInterceptor;
import retrofit.RestAdapter;
import retrofit.http.GET;

/**
 * Meh
 * Created by John on 4/17/2015.
 */
public class MehClient {
    //Replace with your API_KEY
    private static final String API_KEY_VALUE = BuildConfig.MEH_API_KEY;
    private static final String API_URL = "https://api.meh.com/1";
    private static final String PARAM_API_KEY = "apikey";

    private static Meh mMeh;

    public interface Meh {
        @GET("/current.json")
        void getMeh(
                Callback<MehResponse> responseCallback
        );
    }

    public static Meh instance() {
        if (mMeh == null) {
            RestAdapter restAdapter = new RestAdapter.Builder()
                    .setEndpoint(API_URL)
                    .setLogLevel(BuildConfig.DEBUG ? RestAdapter.LogLevel.FULL : RestAdapter.LogLevel.NONE)
                    .setRequestInterceptor(new GiantbombRequestInterceptor())
                    .build();
            mMeh = restAdapter.create(Meh.class);
        }
        return mMeh;
    }

    public static class GiantbombRequestInterceptor implements RequestInterceptor {
        @Override
        public void intercept(RequestFacade request) {
            request.addQueryParam(PARAM_API_KEY, API_KEY_VALUE);
        }
    }
}

File: shared/src/main/java/com/jawnnypoo/openmeh/shared/api/MehClient.kt
package com.jawnnypoo.openmeh.shared.api

import com.jawnnypoo.openmeh.shared.response.MehResponse
import io.reactivex.Single
import okhttp3.OkHttpClient
import okhttp3.logging.HttpLoggingInterceptor
import retrofit2.Retrofit
import retrofit2.adapter.rxjava2.RxJava2CallAdapterFactory
import retrofit2.converter.moshi.MoshiConverterFactory
import retrofit2.http.GET

/**
 * meh.com
 */
class MehClient private constructor() {

    companion object {

        const val API_URL = "https://api.meh.com/"
        const val PARAM_API_KEY = "apikey"

        fun create(apiKey: String, debug: Boolean): MehClient {
            val client = MehClient()
            val clientBuilder = OkHttpClient.Builder()
            if (debug) {
                clientBuilder.addInterceptor(HttpLoggingInterceptor().setLevel(HttpLoggingInterceptor.Level.BODY))
            }
            clientBuilder.addInterceptor { chain ->
                val url = chain.request().url()
                        .newBuilder()
                        .addQueryParameter(PARAM_API_KEY, apiKey)
                        .build()
                val request = chain.request().newBuilder()
                        .url(url)
                        .build()
                chain.proceed(request)
            }
            val restAdapter = Retrofit.Builder()
                    .baseUrl(API_URL)
                    .addConverterFactory(MoshiConverterFactory.create())
                    .addCallAdapterFactory(RxJava2CallAdapterFactory.create())
                    .client(clientBuilder.build())
                    .build()
            client.mehService = restAdapter.create(MehService::class.java)
            return client
        }
    }

    interface MehService {
        @GET("1/current.json")
        fun getMeh(): Single<MehResponse>
    }

    private lateinit var mehService: MehService

    fun getMeh(): Single<MehResponse> {
        return mehService.getMeh()
    }
}
